### DISCUSSION TITLE
Training for detecting and classifying Subclauses and overall Clauses in Developmental Assessments

### DISCUSSION THREAD

**Discussion by Romanang** [explosion/spaCy#12564] [Help: Coding & Implementations] [answered]
Hallo,
 
I am currently investigating into the possibility to automatize a psych-linguistic model I created over the last decade or so. It is meant to measure stages of development based on syntactical complexity and grammar (especially the way subclauses are connected). I already tried spacy which seems to be quite capable of basic operations that are necessary to assess a clause and assign it a likely stage – however, what I would actually need is
(a) a detection of subclauses, i.e., each including subject, predicate, and object, then counting them and assigning a stage based on the number of clauses present and
(b) a classification of conjunctions between subclauses.
 
Here would be an example for (a) – It would need the capacity to figure out that the clause in _File 1_ has three subclauses; 
![File 1](https://user-images.githubusercontent.com/131519356/233772184-1c2cb85f-fa66-4a9a-bf88-06023ee4888b.png)
Versus the following clause having five subclauses in _File 2_
![File 2](https://user-images.githubusercontent.com/131519356/233772185-440e4a87-eddc-4d2b-8e55-ea929fbf9b4f.png)

Could I train that via “span categorization” or could I even try to teach it the classification of the whole clause vs. classifying subclauses and then putting an “if.. then” function over it e.g., “if three subclause then category x”
 
For (b) would dependencies and relations work? Or “named entities” e.g., causal connective vs. adversative connective? Would a word vector help to understand whether subclauses are additive or correlated as in _File 3_ compared to _File 4_

![file 3](https://user-images.githubusercontent.com/131519356/233772186-977551da-3515-4a11-a59a-87839dee4a51.png)
![file 4](https://user-images.githubusercontent.com/131519356/233772187-e1ba5108-4713-4047-89eb-e8aa807b826c.png)
 
I would be happy if you could give me some information on how likely you see this work in prodigy and spacy
Greetings
Roman Angerer

**Comment by rmitsch** [comment, marked as answer]
Hi @Romanang!  Neither clause detection nor classification of conjuctions between subclauses are supported by spaCy out of the box. 

> (a) a detection of subclauses, i.e., each including subject, predicate, and object, then counting them and assigning a stage based on the number of clauses present

For clause detection I recommend having a look at https://spacy.io/universe/project/spacy-clausie/. I think you should be able to build on that to extract your subclauses - the counting step once you've done that should be trivial.

> (b) a classification of conjunctions between subclauses.

Hm, so usual word embeddings aren't very useful for this task, as they don't incorporate any syntactic information. spaCy uses [sense2vec](https://github.com/explosion/sense2vec) for its default word embeddings, which results in richer representations when it comes to that, but even those don't explicitly embed the information that would be helpful to you.
I could imagine that classifying the conjunction by looking at information on the adjacent pair of clauses could work to some degree, but I'm afraid you'll have to roll your own model here - and I'm not sure about the features you'd need to leverage for this.
